fuel-relayer/src/relayer.rs

```diff
@@ -0,0 +1,726 @@
+                    .append_eth_events(&fuel_event, eth_event.block_number.unwrap().as_u64())
+                    .await;
+            }
+
+            // 3. Start listening to eth events
+            let eth_log_filter = Filter::new().address(ValueOrArray::Array(contracts.clone()));
+            watchers = Some((
+                self.stop_handle(|| provider.watch_blocks()).await??,
+                self.stop_handle(|| provider.watch(&eth_log_filter))
+                    .await??,
+            ));
+
+            //let t = watcher.as_mut().expect(()).next().await;
+            // sleep for 50ms just to be sure that our watcher is registered and started receiving events
+            tokio::time::sleep(Duration::from_millis(50)).await;
+
+            // 4. Check if our LastIncludedEthBlock is same as BestEthBlock
+            if best_block == provider.get_block_number().await?
+                && best_block_hash
+                    == self
+                        .stop_handle(|| provider.get_block(best_block))
+                        .await??
+                        .ok_or(RelayerError::InitialSyncAskedForUnknownBlock)?
+                        .hash
+                        .unwrap()
+            {
+                // block number and hash are same as before starting watcher over logs.
+                // we are safe to continue.
+                break;
+            }
+            // If not the same, stop listening to events and do 2,3,4 steps again.
+            // empty pending and do overlaping sync again.
+            // Assume this will not happen very often.
+            self.pending.clear();
+        }
+
+        // 5. Continue to active listen on eth events. and prune(commit to db) dequeue for older finalized events
+        let finalized_da_height = best_block.as_u64() - self.config.eth_finality_slider();
+        self.pending
+            .flush_validator_diffs(&mut *self.db.lock().await, finalized_da_height)
+            .await;
+
+        watchers.ok_or_else(|| RelayerError::ProviderError.into())
+    }
+
+    /// Starting point of relayer
+    #[tracing::instrument(name = "relayer_run", skip_all)]
+    pub async fn run<P>(mut self, provider: P)
+    where
+        P: Middleware<Error = ProviderError>,
+    {
+        //let mut this = self;
+        self.current_validator_set
+            .load_get_validators(&*self.db.lock().await)
+            .await;
+
+        loop {
+            // initial sync
+            let (mut blocks_watcher, mut logs_watcher) = match self.initial_sync(&provider).await {
+                Ok(watcher) => watcher,
+                Err(err) => {
+                    if self.status == RelayerStatus::Stop {
+                        return;
+                    }
+                    error!("Error happened while doing initial sync:{:?}", err);
+                    continue;
+                }
+            };
+            info!("Initial syncing finished. Continue to active sync.");
+            loop {
```

Comment:
does this loop need to be nested?

Reply:
Idea (not fully realized) is to have continue with a loop label so that we can jump if connection with eth client breaks but yea at current setup nesting is not needed.